### DISCUSSION TITLE
Bi-level optimization implementation issues

### DISCUSSION THREAD

**Discussion by MichaelMMeskhi** [google/flax#2695] [Q&A] [answered]
I am trying to implement a similar nested optimization as MAML but am having issues with NaNs. When doing:

```python
grad_fn = jax.value_and_grad(loss_fn)
_, _ = grad_fn(outer.params, inner) # this is a state as I do updates on inner model
```

My error as I debugged I think is due to passing in inner state rather than inner.params. How to overcome this?

**Comment by marcvanzee** [comment]
What is `inner`'s datatype? Generally, you should pass pure functions to JAX transformations, and if you are updating `inner` it seems like it isn't a pure function.

**Comment by cgarciae** [comment, marked as answer]
@MichaelMMeskhi can you post a minimal reproducible example?

**Reply by MichaelMMeskhi** [reply]
I hope [this notebook](https://colab.research.google.com/drive/1gG69i5lJPrrEI8x4MOVy0k02TEWHSRVM?usp=sharing) can be useful.

**Reply by cgarciae** [reply]
Sadly, it seems to be incomplete, can you add the relevant code to run training?

**Reply by MichaelMMeskhi** [reply]
My apologies, it has been updated.

**Reply by cgarciae** [reply]
Seems you managed to solve the problem?

**Reply by MichaelMMeskhi** [reply]
I am not quite sure. it seems to work but I am trying to validate the implementation. Essentially, I am trying to update an outer model's parameters based on minimizing the loss of an inner model. The issue I used to encounter was if I have some function that I try to compute the gradient of say `loss_fn(outer_model.params, inner_model, ...)`, I believe passing `inner_model` causes `nans` as this is a state that is further optimizing internally. Then the loss of that optimized `inner_model` should be used to optimizing `outer_model`. Sorry if this doesn't make sense I am just trying to figure out how to correctly implement all of this in Jax. Because later doing `jax.value_and_grad(loss_fn)(outer_model.params, inner_model, ...)` causes the issue.

**Reply by cgarciae** [reply]
I don't have experience with this topic specifically, but it seems you want to actually update the inner state in the outer loop:

```diff
        for _ in range(1): # k-steps of gradient updates
            inner_grad_fn = jax.value_and_grad(inner_loss_fn, has_aux=False)
            loss, inner_grads = inner_grad_fn(inner_state.params)
-            updated_inner_state = inner_state.apply_gradients(grads= inner_grads)
+            inner_state = inner_state.apply_gradients(grads= inner_grads)

-        return loss, updated_inner_state
+        return loss, inner_state
```
Else you are not emulating gradient updates.

**Reply by MichaelMMeskhi** [reply]
I found my issue to cause `nan` gradients coming from `jnp.multiply(nll, z)`. Here `z` is some regularization term I use to scale the output. Does using `jnp.multiply` detach gradients from the graph hence causing `nan`?
